go/database/mpt/nodes_test.go

```diff
@@ -6869,6 +6869,49 @@ func TestVisitPathToStorage_CanReachTerminalNodes(t *testing.T) {
 	}
 }
 
+func TestVisitPathToStorageWithHashAccess_Nodes_Hash_Protected(t *testing.T) {
+	desc := &Extension{path: []Nibble{1, 2}, next: &Branch{children: Children{
+		0x4: &Account{
+			address: common.Address{0x12, 0x42},
+			info:    AccountInfo{Nonce: common.Nonce{1}},
+			storage: &Value{key: common.Key{0x12}, value: common.Value{1}},
+		},
+	}}}
+
+	ctrl := gomock.NewController(t)
+	ctxt := newNiceNodeContext(t, ctrl)
+
+	var pathFound bool
+	root, _ := ctxt.Build(desc)
+	if _, err := visitPathToAccountWithHashAccess(ctxt, &root, common.Address{0x12, 0x42}, MakeVisitor(func(node Node, info NodeInfo) VisitResponse {
+		if account, ok := node.(*AccountNode); ok {
+			if _, err := visitPathToStorageWithHashAccess(ctxt, &account.storage, common.Key{0x12}, MakeVisitor(func(storageNode Node, storageInfo NodeInfo) VisitResponse {
+				// node cannot be accessed via view access as long as hash access is held
+				if _, ok := ctxt.getHandle(t, storageInfo.Id).TryGetViewHandle(); ok {
+					t.Errorf("expected node to be hash protected, but it is not")
+				}
+				pathFound = true
+				return VisitResponseContinue
+			})); err != nil {
+				t.Fatalf("unexpected error during visit: %v", err)
+			}
+			return VisitResponseContinue
+		}
+		
+		// node cannot be accessed via view access as long as hash access is held
+		if _, ok := ctxt.getHandle(t, info.Id).TryGetViewHandle(); ok {
+			t.Errorf("expected node to be hash protected, but it is not")
+		}
+		return VisitResponseContinue
+	})); err != nil {
```

Comment:
This is a very long condition with plenty of side-effects. As mentioned in another PR review, this is not easy to parse while reviewing it.

This condition also covers tests for two different functions - the path-to-account and path-to-storage. Splitting these up into two individual tests may improve readability.